### PR TITLE
Geopackage: Add upsert operation support

### PR DESCRIPTION
**Pull request by beryan**
## What does this PR do?
This PR implements the Upsert feature operation in the OGR Geopackage driver (ie. if record does not exist then Create, else Set).

## What are related issues/pull requests?
For context, the OGR Upsert operation was added here https://github.com/OSGeo/gdal/pull/6199

## Tasklist

 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp [hunk 1]
```diff
@@ -2039,6 +2039,26 @@ OGRErr OGRGeoPackageTableLayer::ISetFeature( OGRFeature *poFeature )
...
+/************************************************************************/
+
+OGRErr OGRGeoPackageTableLayer::IUpsertFeature( OGRFeature* poFeature )
+
+{
+   if ( !TestCapability(OLCUpsertFeature) )
+      return OGRERR_FAILURE;
+
+   if ( GetFeature(poFeature->GetFID()) )
+   {
```
[the remark this answers is not in the document]
Comment (beryan) [reply ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp:2083]: This is true. Based on create/set feature implementation complexity, which has additional logic that differs between the functions beyond the SQL syntax, I am more inclined to redirect and fix the memory concern.

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp [hunk 2]
```diff
@@ -2039,6 +2039,56 @@ OGRErr OGRGeoPackageTableLayer::ISetFeature( OGRFeature *poFeature )
...
+/************************************************************************/
+/*                          FeatureIDExists()                           */
+/************************************************************************/
+
+bool OGRGeoPackageTableLayer::FeatureIDExists( GIntBig nFID )
+{
+   CPLString soSQL;
+   soSQL.Printf("SELECT %s FROM \"%s\" m "
+                "WHERE \"%s\" = " CPL_FRMT_GIB,
+                m_soColumns.c_str(),
```
Comment (rouault) [inline ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp:2051]: could be simplified as:
```suggestion
   soSQL.Printf("SELECT 1 FROM \"%s\" "
                "WHERE \"%s\" = " CPL_FRMT_GIB,
```

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp [hunk 3]
```diff
@@ -2039,6 +2039,56 @@ OGRErr OGRGeoPackageTableLayer::ISetFeature( OGRFeature *poFeature )
...
+
+   sqlite3_stmt* poStmt = nullptr;
+   int err = sqlite3_prepare_v2(
+      m_poDS->GetDB(), soSQL.c_str(), -1, &poStmt, nullptr);
+   if ( err != SQLITE_OK )
+   {
+      sqlite3_finalize(poStmt);
+      CPLError(CE_Failure, CPLE_AppDefined,
+               "failed to prepare SQL: %s", soSQL.c_str());
+      return nullptr;
```
Comment (rouault) [inline ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp:2064]:
```suggestion
      return false;
```

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rouault) [reply ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp:2083]: fair point

**Pull request merged**